library/Fission/User/Modifier.hs

```diff
@@ -36,13 +42,65 @@ updatePublicKeyDB ::
   => UserId
   -> Key.Public
   -> UTCTime
-  -> Transaction m ()
-updatePublicKeyDB userID pk now =
+  -> Transaction m (Either Errors Key.Public)
+updatePublicKeyDB userID pk now = do
   update userID
     [ UserPublicKey  =. Just pk
     , UserModifiedAt =. now
     ]
-  
+
+  return $ Right pk
+
+addExchangeKeyDB ::
+     MonadIO m
+  => UserId
+  -> RSA.PublicKey
+  -> UTCTime
+  -> Transaction m (Either Errors [RSA.PublicKey])
+addExchangeKeyDB userID key now =
+  Persist.get userID >>= \case
+    Nothing -> 
+      return . openLeft $ NotFound @User
+
+    Just user -> do
+      let keys = userExchangeKeys user
+      if List.elem key keys
+        then 
+          return $ Right keys
+
+        else do
+          let updated = keys ++ [key]
```

Comment:
This ordering is actually pretty inefficient, because lists are linked lists and here the innermost one is being changes. Two ways to make this faster:

# 1. Reverse the Ordering

```haskell
let updated = key : keys
```

2. Use a [Vector](https://www.stackage.org/haddock/lts-15.5/rio-0.1.14.1/RIO-Vector.html) [essentially an array]

```haskell
import RIO.Vector
-- ...
let updated = keys ++ singleton key
```

(where `++` is from `RIO.Vector`)

Reply:
ah cool cool :+1: